commcare_connect/events/models.py

```diff
@@ -0,0 +1,144 @@
+from abc import ABCMeta, abstractproperty
+from dataclasses import dataclass
+from datetime import datetime
+
+from django.db import models
+from django.utils.translation import gettext as _
+
+from commcare_connect.users.models import User
+
+from . import types
+
+
+class Event(models.Model):
+    from commcare_connect.opportunity.models import Opportunity
+
+    # this allows referring to event types in this style: Event.Type.INVITE_SENT
+    Type = types
+
+    date_created = models.DateTimeField(auto_now_add=True, db_index=True)
+    event_type = models.CharField(max_length=40, choices=types.EVENT_TYPE_CHOICES)
```

Comment:
You mentioned you were going to rethink this in response to my last comment but appears to still be here? How has your thinking here evolved? Why did you decide to require a migration each time we have a new event?

Reply:
Yeah, I am going to make it a [callable](https://github.com/django/django/commit/691f70c47755c7f55fa75ce607d028f81468b745) to avoid migrations getting created

Reply:
So, it looks like this won't work since we are on Django 4 (and callable only works in 5), we can do it using [validators](https://docs.djangoproject.com/en/4.2/ref/validators/). But the trade off is that we need to have some extra code to toggle the slug and verbose_name (and back) while processing in some of the places (which is ugly and unnecessary). I could live with extra migration until Django 5 instead of the unnecessary code. Do you feel strongly against migrations?

Reply:
> Do you feel strongly against migrations?

I am not sure I totally understand why we need restricted choices here at all. That seems pretty nonstandard for an analytics system (if you think of GA or kissmetrics, they allow arbitrary events), and makes adding new events, whcih we expect to do incrementally much trickier. Mobile will need to coordinate releases around web running these migrations, and simple communication errors could cause confusing problems. I like the idea on web of predefining our events, and only using ones we have listed in something like a constant file, but I am not sure I am sold on enforcing them at the db level.

As with other comments, I am open to being convinced